### PR TITLE
Chef extension ARM commands

### PR DESCRIPTION
**Pull request by NimishaS**
1. Using new classes like `CreateOrUpdateWithHttpMessagesAsync` for `Set-AzureRmVMChefExtension`
2. Passing settings file as Hashtable
3. Implements `Get-AzureRmVMChefExtension`
4. Implements `Remove-AzureRmVMChefExtension`

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@NimishaS__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

**Comment by NimishaS** [comment]
@markcowl , Travis is failing but I don't have the permission to see the details. Please help.

**Comment by hovsepm** [comment]
Several issues:
1. Help missing for cmdlet Get-AzureRmVMChefExtension implemented by class Microsoft.Azure.Commands.Compute.Extension.Chef.GetAzureRmVMChefExtension 
2. Help missing for cmdlet Remove-AzureRmVMChefExtension implemented by class Microsoft.Azure.Commands.Compute.Extension.Chef.RemoveAzureRmVMChefExtension

File: src/ResourceManager/Compute/Commands.Compute/Extension/Chef/GetAzureRmVMChefExtension.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
+﻿using System;
```
Comment (hovsepm) [inline src/ResourceManager/Compute/Commands.Compute/Extension/Chef/GetAzureRmVMChefExtension.cs]: Please add License header

File: src/ResourceManager/Compute/Commands.Compute/Extension/Chef/RemoveAzureRmVMChefExtension.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
+﻿using System;
```
Comment (hovsepm) [inline src/ResourceManager/Compute/Commands.Compute/Extension/Chef/RemoveAzureRmVMChefExtension.cs]: License header missing

**Comment by hovsepm** [comment]
@NimishaS please add test coverage.

**Comment by NimishaS** [comment]
@hovsepm we faced some issues in past while running the tests in `recording` mode. We had followed the documentation: https://github.com/Azure/azure-powershell/wiki/Microsoft-Azure-PowerShell-Developer-Guide#ad-scenario-tests, but couldn't get pass through.

Is it possible for you to schedule an interactive session to help us guide how the test recording can be done? We'll keep the specs ready.

This is the PR which couldn't be merged because of test recording issue: https://github.com/Azure/azure-powershell/pull/1676

**Comment by NimishaS** [comment]
@hovsepm , any updates here?

**Comment by hovsepm** [comment]
@NimishaS what I understand from you previous PR is that your tenant ID does not have any subscription. To be able to help you I'll need access to your test environment and devbox. Do you have access to MS corp net?

**Comment by markcowl** [comment]
@azuresdkci retest this please

**Comment by NimishaS** [comment]
@hovsepm we don't have access to MS corp net. I can share the details of a VM which will have our setup. If this approach sounds fine then please share your email id on `nimisha.sharad@msystechnologies.com`.

@markcowl , @hovsepm : Do you expect us to add test suite in this PR only or can we add that in a separate PR? Finishing test suite will take some time as we need to co-ordinate with you for that.

**Comment by markcowl** [comment]
@NimishaS As long as you have tested the cmdlets thoroughly, you commit to testing them in the candidate installer during the release, and you file an issue to add the test suite (and make sure to do out before the end of next sprint), then we can defer this.  If you decide to go this route, please add a reference to the github issue for adding the tests here.

**Comment by NimishaS** [comment]
@markcowl , created https://github.com/Azure/azure-powershell/issues/2638. Please assign this issue to me. I don't have the rights.
We have tested the cmdlets thoroughly and will also take up the task to test them during the release on high priority. Since we are ruby developers and new to C# and visual studio, it's not clear to me what it means to test in candidate installer. Can you please share some links?

File: src/ResourceManager/Compute/Commands.Compute/Extension/Chef/RemoveAzureRmVMChefExtension.cs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+using System.Globalization;
+using AutoMapper;
+using Microsoft.Azure.Commands.Compute.Models;
+using Microsoft.Azure.Management.Compute.Models;
+
+namespace Microsoft.Azure.Commands.Compute.Extension.Chef
+{
+    [Cmdlet(
+        VerbsCommon.Remove,
+        ProfileNouns.VirtualMachineChefExtension)]
```
Comment (markcowl) [inline src/ResourceManager/Compute/Commands.Compute/Extension/Chef/RemoveAzureRmVMChefExtension.cs]: SupportsShouldProcess=true

File: src/ResourceManager/Compute/Commands.Compute/Extension/Chef/RemoveAzureRmVMChefExtension.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            if (this.Linux.IsPresent)
+            {
+                this.Name = LinuxExtensionName;
+            }
+            else if (this.Windows.IsPresent)
+            {
+                this.Name = ExtensionDefaultName;
+            }
+
+            ExecuteClientAction(() =>
```
Comment (markcowl) [inline src/ResourceManager/Compute/Commands.Compute/Extension/Chef/RemoveAzureRmVMChefExtension.cs]: Protect with PSCmdlet.ConfirmAction(target, processMessage, Action)

File: src/ResourceManager/Compute/Commands.Compute/Extension/Chef/SetAzureVMChefExtension.cs [hunk 5]
```diff
@@ -1,19 +1,33 @@
...
+using System.Collections;
 using System.IO;
+using System.Linq;
+using System.Management.Automation;
 
 namespace Microsoft.Azure.Commands.Compute.Extension.Chef
 {
     [Cmdlet(
         VerbsCommon.Set,
         ProfileNouns.VirtualMachineChefExtension)]
```
Comment (markcowl) [inline src/ResourceManager/Compute/Commands.Compute/Extension/Chef/SetAzureVMChefExtension.cs]: Add SupportsShouldProcess = true

**Comment by markcowl** [comment]
@NimishaS,  A couple of additoonal comments.  Please see: https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d   For more information on implementing the ShouldProcess pattern using ConfirmAction.  We are now requiring this on all new cmdlets (and wnever a cmdlet is changed)

**Comment by markcowl** [comment]
@NimishaS  Please update per the review comments.   Your invitation is here:  https://github.com/Azure/azure-powershell/invitations  this will enable you to assign yourself to the issue (and close when it is done)

**Comment by NimishaS** [comment]
Sorry @markcowl, I had got busy with knife-ec2 release stuff.I will update for the review comments.

**Comment by NimishaS** [comment]
@markcowl , I have got a little occupied with `knife-ec2` release. I will be picking this up again in 2-3 days, once `knife-ec2` is out.

**Comment by NimishaS** [comment]
@markcowl , added `SupportsShouldProcess` and `ConfirmAction` for `Set-AzureRmVMChefExtension` and `Remove-AzureRmVMChefExtension` commands. Also tested with `-confirm` option.

**Comment by NimishaS** [comment]
@markcowl I will be able to start with https://github.com/Azure/azure-powershell/issues/2638 once this PR is merged as I need to create the PR for https://github.com/Azure/azure-powershell/issues/2638 from `dev` branch. Please review this and let us know if anything needs to be fixed.

**Comment by markcowl** [comment]
on demand run here: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1032/

LGTM once the on-demand run passes

**Pull request merged**